### ISSUE TITLE
Memory leak 

### ISSUE THREAD

**Issue by MProx** [tensorflow/tensorflow#33009]
**System information**
**- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):** Yes, see below
**- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):** Ubuntu 18.04 (also demonstrated in Windows 7)
**- TensorFlow installed from (source or binary):** Binary
**- TensorFlow version (use command below):** v2.0.0-rc2-26-g64c3d38 2.0.0 (problem disappears on v1.14.0-rc1-22-gaf24dc91b5 1.14.0)
**- Python version:** 3.6.8
**- CUDA/cuDNN version:** Cuda v10.0, CuDNN v7.6.2.24
**- GPU model and memory:** Nvidia GeForce 840M, but problem persists in non-GPU version

**Describe the current behavior**
When creating a trivially simple model and then entering a loop that calls predict() with dummy input, memory consumption increases indefinitely over time. On my system, a model with a single hidden layer of only 32 nodes will consume all available system RAM (>10gb) after only 10 minutes. The problem happens on v2.0 (GPU or CPU) of tensorflow, but disappears when running identical code on v1.14.

**Describe the expected behavior**
Expect memory consumption to quickly stabilize but it never does.

**Code to reproduce the issue**

```
from tensorflow.keras import Input, Model
from tensorflow.keras.layers import Dense
import numpy as np

# Build model
In = Input(shape=(10,))
x = Dense(32)(In)
Out = Dense(2)(x)

# Compile
model = Model(inputs=In, outputs=Out)
model.compile(optimizer='adam', loss='mse')        

# Create dummy input data
fake_data = np.random.uniform(low=0, high=1.0, size=(1, 10, ))

while True:
    # Repeatedly predict:
    model.predict(fake_data) # No memory leak if this line is replaced with "pass"
```

**Comment by ialdencoots** [comment]
I have also experienced this critical issue.

**Comment by MProx** [comment]
I have managed to get around this error by using `model.predict_on_batch()` instead of `model.predict()`. This returns an object of type `<class 'tensorflow.python.framework.ops.EagerTensor'>` - not a numpy array as claimed in [the docs](https://www.tensorflow.org/api_docs/python/tf/keras/Model#predict_on_batch) - but it can be cast by calling `np.array(model.predict_on_batch(input_data))` to get the output I want.

I also noticed a similar memory leak problem with calling `model.fit()` in a loop, albeit with a slower memory accumulation, but this can be fixed in a similar way using `model.train_on_batch()`.

**Comment by ipsec** [comment]
I think same error here.
This code (https://github.com/ipsec/dqn_tf2/blob/master/nchain-train.py) not finishing.

**Comment by oanush** [comment]
@MProx ,
Can you confirm if the issue is resolved with the workaround? Thanks!

**Comment by MProx** [comment]
Hi. Yes it did. I tested it overnight with no leak. I'm not sure why the original functions do not work as expected, but the workaround has fixed my problem.

**Comment by ialdencoots** [comment]
This issue is NOT resolved. I use predict_on_batch and still get an OOM error after some time of processing data.

**Comment by MProx** [comment]
Interesting. I think that OOM is telling you your GPU memory is being depleted, and that sounds like a different problem because mine was filling up system RAM, not GPU. Using predict_on_batch() and train_on_batch(), I ran ~800k iterations of my model over 10 hours last night, and it seemed to work fine.

Another thing is that I had to set the environment variable TF_FORCE_GPU_ALLOW_GROWTH=true to avoid tensorflow errors on startup. Maybe try that?

**Comment by ialdencoots** [comment]
Yeah, I don't have any critical errors when running on CPU. It's definitely GPU memory that's leaky. The GPU memory doesn't deplete immediately, but only after some number of calls to `predict_on_batch`. I am able to train using `fit` with no problems, which runs over that same data set over many epochs.

**Comment by ialdencoots** [comment]
I can confirm actually, that when run on CPU, the leak still persists.

**Comment by MProx** [comment]
Hmm, in that case I'm

**Comment by MProx** [comment]
In that case I'm going to re-open this issue so that maybe the TF team can help you.

**Comment by oanush** [comment]
Issue replicating for TF-2.0, kindly find the [gist](https://colab.sandbox.google.com/gist/oanush/07da0eef3fbfbd8ff74e6c840ad5b6d5/33009.ipynb) of colab.Thanks!

**Comment by robieta** [comment]
This is an issue with how certain object lifetimes are being managed in the tf function cache. Behind the scenes `Model.predict` is creating some functions which aren't being spun down properly, which is why there is a small but consistent leak each iteration. We are currently working on a fix.

**Comment by novog** [comment]
@robieta, would the particular problem that you mentioned affect GPU memory, or just system memory? Is there a Github Issue associated with it?

**Comment by novog** [comment]
I was also running a `predict` loop and getting OOM errors after many iterations. Using `predict_on_batch` instead eliminated this problem. It also sped up the prediction rate by a factor of 4. Thanks for the workaround!

@ialdencoots, could you confirm whether or not you get an OOM error using the specific example code that @MProx provided in the issue description except using `predict_on_batch`?

**Comment by ialdencoots** [comment]
@novog Hmm, using the code that MProx provided with `predict_on_batch`, I do not get the OOM error. In my considerably less minimal code, where I am feeding inputs via a `TFRecordDataset` and definitely using `predict_on_batch`, I still encounter the memory leak.

**Comment by novog** [comment]
@ialdencoots It sounds like you might be having a different OOM issue than the author of this issue and I were having. Incidentally, in my case, I am also loading my model from an .h5 file, and `predict_on_batch` fixed it. I would suggest trying to create a simpler version of your code that still has the problem to narrow down the cause and post it as a separate issue.

@MProx, I would recommend leaving this issue open. Although we have a workaround, there is a defect in the TensorFlow version of Keras's `predict` function that should be fixed.

**Comment by Exxion** [comment]
In case it helps, this issue does not appear to be present in release 2.0.0.alpha0. You can confirm this by running @oanush's Colab with that version. (Not that it's relevant to this issue, but the same is also true of every other issue I've encountered when working with TensorFlow 2.0.)

**Comment by mihaimaruseac** [comment]
Please switch to 2.0.0 (final release). You might need to upgrade `pip` and `setuptools` to get it, but 2.0 has been available for a while, no need to still be on alpha

**Comment by Exxion** [comment]
I know. But this problem, as well as several others, is not present on alpha.

**Comment by mihaimaruseac** [comment]
Oh, you meant that the breakage happens in between alpha and rc2.

If it's not too much to ask, would you be willing to compare rc1?

**Comment by Exxion** [comment]
The issue does appear to be present in rc1.
I went ahead and tried beta0 as well, and it does *not* seem to be present there.

**Comment by mihaimaruseac** [comment]
Thank you. so the failure is in between beta1 and rc0.

Unfortunately, we fast-forwarded the branch in between these two points, so probably we got the failure from master.

One more test, please. Is the issue also manifesting on latest master?

**Comment by mihaimaruseac** [comment]
Oh, I see @robieta already [pinned the issue](https://github.com/tensorflow/tensorflow/issues/33009#issuecomment-539226841). Apologies for the extra tests, we are expecting this to be fixed on master soon (if not already)

**Comment by Exxion** [comment]
No worries about the tests. I just hope a new release with the fix comes out soon.

**Comment by robieta** [comment]
It's worth noting that the solution that I was referring to will only fix CPU OOM; it wouldn't explain a GPU leak.

**Comment by williamccrobinson** [comment]
I was experiencing the same issue on the 2.0.0 docker (latest-gpu-py3) image; I changed to the nightly image (tensorflow/tensorflow:nightly-gpu-py3 to be exact) and the problem seems to be fixed, no more memory leak!

**Comment by marnixkoops** [comment]
Avoiding `model.predict` and using `model.predict_on_batch` solves the Memory Error for me. Here's an example to create batched predictions on your test set. 
(Note that if only pre-allocating the results array already results in a MemoryError then simply the array does not fit in your available memory regardless of the memory leak issue.)

```
# custom batched prediction loop to avoid memory leak issues for now in the model.predict call
y_pred_probs = np.empty([len(X_test), VOCAB_SIZE], dtype=np.float32)  # pre-allocate required memory for array for efficiency

BATCH_INDICES = np.arange(start=0, stop=len(X_test), step=BATCH_SIZE)  # row indices of batches
BATCH_INDICES = np.append(BATCH_INDICES, len(X_test))  # add final batch_end row

for index in np.arange(len(BATCH_INDICES) - 1):
    batch_start = BATCH_INDICES[index]  # first row of the batch
    batch_end = BATCH_INDICES[index + 1]  # last row of the batch
    y_pred_probs[batch_start:batch_end] = model.predict_on_batch(X_test[batch_start:batch_end])
```

**Comment by sedghi** [comment]
`model.predict_on_batch` solved this problem for me too

**Comment by jc-louis** [comment]
Same issue with TF 2.0 stable
Solved with `tf.compat.v1.disable_eager_execution()`

**Comment by Tauranis** [comment]
I got the same problem both on rc2 and on official 2.0.0 release.
`model.predict_on_batch` workaround worked for me

**Comment by JakobGM** [comment]
For anyone visiting this thread with similar issues of memory leakage when invoking `model.evaluate()` in a loop, the work-around for this case is similar to the one proposed by @MProx, use `model.test_on_batch()` instead.

**Comment by Mike575** [comment]
Building a model by subclassing the `Model` class can avoid this problem.

``
class MyModel(tf.keras.Model):
    def __init__(self):
        super(MyModel, self).__init__()
        self.dense1 = tf.keras.layers.Dense(32, activation=tf.nn.relu)
        self.dense2 = tf.keras.layers.Dense(2, activation=tf.nn.softmax)

    def call(self, inputs):
        x = self.dense1(inputs)
        return self.dense2(x)

model = MyModel()
model.compile(optimizer='adam', loss='mse')        

# Create dummy input data
fake_data = np.random.uniform(low=0, high=1.0, size=(1, 10, ))

while True:
    # Repeatedly predict:
    model.predict(fake_data) # memory leak will happen
    model(fake_data) # There is no memory leak
'

**Comment by rchao** [comment]
@MProx, about a month ago we added a fix for memory leak and there's a possibility it has fixed this. Can you try `!pip install tf-nightly` and see if it resolves your issue?

**Comment by MProx** [comment]
Hi. I have done what you ask and tested `tf-nightly` with my trivial code example from the original issue above, and it does indeed seem to fix the problem. I also tried it in a different application that I've been using the predict_on_batch workaround with, and there seem to be no memory leaks there either. So yes, the leak seems to have been dealt with. 
Thank you!

**Comment by rchao** [comment]
Thanks @MProx for the updates!

**Comment by yashmanuda** [comment]
Hi, 

I am trying to call `model.predict()` on CPU (model trained on GPU) multiple times and I observe RAM memory leak. `clear_session` with model reload doesn't solve the issue. `model.predict_on_batch()` fails to solve the issue as well. Is there a workaround for this issue? I am using TF 1.13 and Python 3.6. Have been struggling to solve this problem since so long. Kinda need help.

**Comment by MProx** [comment]
Does the issue persist of you use the minimal code example I supplied in
the original issue post?

**Comment by yashmanuda** [comment]
Yes, the original code as shown below causes memory leak.

```
from tensorflow.keras import Input, Model
from tensorflow.keras.layers import Dense
import numpy as np

# Build model
In = Input(shape=(10,))
x = Dense(32)(In)
Out = Dense(2)(x)

# Compile
model = Model(inputs=In, outputs=Out)
model.compile(optimizer='adam', loss='mse')        

# Create dummy input data
fake_data = np.random.uniform(low=0, high=1.0, size=(1, 10, ))

while True:
    # Repeatedly predict:
    model.predict(fake_data) # No memory leak if this line is replaced with "pass"
```

Below is the memory consumption over time (it increases initially which is expected, but after certain point when it's supposed to be constant, it rather increases although by a very small amount ~0.25 MB, I ran it for only 30 minutes)
```
CMDLINE python test.py
MEM 0.609375 1579247671.2321
MEM 22.660156 1579247671.3397
MEM 68.218750 1579247671.4467
MEM 104.500000 1579247671.5552
MEM 114.675781 1579247671.6605
MEM 124.125000 1579247671.7657
MEM 131.191406 1579247671.8710
MEM 146.972656 1579247671.9783
MEM 156.164062 1579247672.0838
MEM 168.886719 1579247672.1893
MEM 182.667969 1579247672.2945
MEM 192.363281 1579247672.4000
MEM 202.539062 1579247672.5052
MEM 211.316406 1579247672.6102
MEM 211.968750 1579247672.7151
MEM 215.250000 1579247672.8199
MEM 215.464844 1579247672.9256
MEM 215.738281 1579247673.0316
MEM 215.890625 1579247673.1366
MEM 216.050781 1579247673.2414
MEM 216.468750 1579247673.3461
MEM 221.535156 1579247673.4508
MEM 226.820312 1579247673.5557
MEM 227.011719 1579247673.7651
MEM 227.261719 1579248604.1910
MEM 227.511719 1579249397.2364
```

**Comment by nsarang** [comment]
I've had the same problem and none of the above solutions worked for me.
Ultimately, setting `inter_op_parallelism_threads=1` in `tf.Config` fixed it.
tensorflow/tensorflow#22098

**Comment by vizaiapp** [comment]
Still getting this error with tf-nightly

**Comment by bentyeh** [comment]
@rchao @MProx Still getting this error with TensorFlow version 2.2.0-rc3 on Google Colab. See this [gist](https://colab.research.google.com/gist/bentyeh/5a624a95f8614a7d140592e62b7967fa/tf_keras_issue33009.ipynb).

**Comment by rajanlagah** [comment]
Same getting error even after using `train_on_batch` i am using custom keras layers

**Comment by nitinmnsn** [comment]
The issue persists. Would be better if someone from tensorflow team made a statement about it. It severely hinders the ability to train for higher number of epochs

**Comment by pallavisrivastava123** [comment]
Hi,

I am trying to call `model.predict()` on CPU multiple times and I observe RAM memory leak. clear_session with model reload and gc.collect() doesn't solve the issue. I ran the code on tensorflow 2.1 and 2.3 as well but issue still persists. Is there a workaround for this issue? I am using TF 1.14 and Python 3.6. Have been struggling to solve this problem since so long.

**Comment by atyshka** [comment]
Having similar issue in latest TF 2.4.1. Growing from about 60GB (I'm using a large shuffle buffer) to 128GB over the course of a few hours. Not sure if it't the same issue as originally mentioned here, since it seems to be more subtle and any number of causes could be the memory leak. Would file a new issue but it's challenging to create a minimal reproducible example

**Comment by Prajval-1608** [comment]
Is the issue still open? I am having similar issue with tf.predict() function on TF 2.11.0.

**Comment by DmytroLopushanskyy** [comment]
Witnessing the same issue while running inference with TF

**Comment by cesar-salcedo** [comment]
Yes, I experience the same issue even with .predict_on_batch. In this case, the memory leak is minor, but it persists.

**Comment by boarder7395** [comment]
Seeing this issue with tensorflow 2.13.1, running python 3.11.8. I ran tracemalloc to determine which objects were not being garbage collected. This is just a snippet of the top offenders if a more complete list is needed I can include that as well.

```tensorflow/python/framework/constant_op.py:98: size=110 MiB, count=1808660, average=64
tensorflow/python/ops/gen_array_ops.py:10241: size=52.2 MiB, count=854768, average=64
tensorflow/python/ops/gen_math_ops.py:6577: size=36.6 MiB, count=600142, average=64
tensorflow/python/ops/gen_math_ops.py:462: size=33.8 MiB, count=553135, average=64
tensorflow/python/ops/gen_math_ops.py:9854: size=31.4 MiB, count=514545, average=64
tensorflow/python/ops/gen_math_ops.py:9738: size=30.1 MiB, count=492770, average=64
tensorflow/python/ops/gen_math_ops.py:6014: size=22.9 MiB, count=375486, average=64
tensorflow/python/ops/gen_math_ops.py:11793: size=20.9 MiB, count=341905, average=64
tensorflow/python/ops/gen_array_ops.py:10697: size=18.3 MiB, count=300004, average=64
tensorflow/python/ops/gen_math_ops.py:5110: size=15.8 MiB, count=258130, average=64
tensorflow/python/ops/gen_nn_ops.py:850: size=13.3 MiB, count=217968, average=64
tensorflow/python/eager/execute.py:53: size=13.1 MiB, count=214666, average=64```

**Comment by boarder7395** [comment]
Also tested tensorflow 2.15.1 and python 3.11.8 this morning with the same results as above

**Comment by boarder7395** [comment]
<script src="https://gist.github.com/boarder7395/d64e372677e8af7d9fdded68eb63f108.js"></script>
I have included the above gist to reproduce the issue. 

Here are the results after 1000 runs on python 3.11.8 with tensorflow 2.15.1
Memory usage after 100 run(s) (in MiB): 839.238
Memory usage after 200 run(s) (in MiB): 843.320
Memory usage after 300 run(s) (in MiB): 847.152
Memory usage after 400 run(s) (in MiB): 851.035
Memory usage after 500 run(s) (in MiB): 854.895
Memory usage after 600 run(s) (in MiB): 858.754
Memory usage after 700 run(s) (in MiB): 862.617
Memory usage after 800 run(s) (in MiB): 866.484
Memory usage after 900 run(s) (in MiB): 870.637
Memory usage after 1000 run(s) (in MiB): 874.645

**Comment by redbaron4** [comment]
@boarder7395 I have run into similar issue with `.fit()` when calling it multiple times (Training multiple models). Have documented findings in [this thread](https://github.com/tensorflow/tensorflow/issues/58676#issuecomment-2060579656)

**Comment by eladmeir** [comment]
I have tried a lot of different TF versions, and most of the available "solutions" on github - and still couldn't stop TF from leaking.
The CPU RAM is always growing, even if I get to lower the rate of growth (by using `predict_on_batch`, which was the only thing that helped, but did not solve)

This thing makes it impossible to maintain a live TF inference server - at some time, if you do not restart the machine, the CPU RAM will overload and the machine would crash

**Comment by cesar-salcedo** [comment]
I gave up on it and ended up exporting the model to .onnx for inference.
